### ISSUE TITLE
Fix configuration comments within selector and value lists being ignored

### ISSUE THREAD

**Issue by fiskhandlarn** [stylelint/stylelint#3111]
I was told in https://github.com/rstacruz/stylelint-rscss/issues/11#issuecomment-350542523 that this is a problem in `stylelint` rather than `stylelint-rscss`.

> Describe the issue. Is it a bug or a feature request (new rule, new option, etc.)?

```
    &:checked ~ .single-steward-header, // stylelint-disable-line rscss/class-format
    &:checked ~ .toggle {
      //
    }
```

still triggers an error for the first line: 
`Invalid element name: '.single-steward-header'   rscss/class-format`

ATM you have to do this instead:
```
    // stylelint-disable
    &:checked ~ .single-steward-header,
    // stylelint-enable
    &:checked ~ .toggle {
      //
    }
```

> Which rule, if any, is this issue related to?

`stylelint-disable-line`

> What CSS is needed to reproduce this issue?

See above.

> What stylelint configuration is needed to reproduce this issue?

```json
{
  "extends": [
    "stylelint-rscss/config"
  ]
}
```

> Which version of stylelint are you using?

8.2.0

> How are you running stylelint: CLI, PostCSS plugin, Node API?

Webpack with `stylelint-webpack-plugin`.

> Does your issue relate to non-standard syntax (e.g. SCSS, nesting, etc.)?

Yes, SCSS.

> What did you expect to happen?

No warnings to be flagged.

> What actually happened (e.g. what warnings or errors you are getting)?

The following warnings were flagged:

```shell
Invalid element name: '.single-steward-header'   rscss/class-format
```

**Comment by jeddy3** [comment]
@fiskhandlarn Thanks for the report and for using the template.

Unfortunately, this is a known limitation. It's document at the bottom of [this section of the configuration guide](https://github.com/stylelint/stylelint/blob/master/docs/user-guide/configuration.md#turning-rules-off-from-within-your-css). "Comments within _selector and value lists_ are currently ignored.". 

Having said that, digging through the history it looks like [it was an issue](https://github.com/stylelint/stylelint/issues/1628) with postcss@5. stylelint now uses postcss@6. Does anyone [know if this behaviour changed](https://github.com/postcss/postcss/issues/845) in postcss@6? If so, can this limitation be addressed now?

**Comment by fiskhandlarn** [comment]
Ah, I missed that, @jeddy3. Thanks for following up anyway. :)

**Comment by jeddy3** [comment]
Closing as a documented limitation.

**Comment by jeddy3** [comment]
Whoops, it seems I was a bit gung-ho in my issue triaging. 

@fiskhandlarn If you'd like to see this issue resolved, please consider investigation whether this can be fixed in stylelint now that we use postcss@6, or whether the limitation is still upstream.

**Comment by fiskhandlarn** [comment]
@jeddy3 Thanks for reopening!

As far as I can see, this package was already using postcss@6 when I opened this issue: https://github.com/stylelint/stylelint/blob/1f611bda663350c99a0893315887418aab8f7095/package.json

I don't know if that's what you meant though. Eitherway, I'm not sure how I can proceed investigating this further since my knowledge about these packages are quite limited, I was hoping you could help. X)

**Comment by fiskhandlarn** [comment]
@jeddy3 Anything I can do to help? :)

**Comment by jeddy3** [comment]
> Anything I can do to help? :)

Thanks for offering to help. I can certainly point you in a direction.

In [`assignDisabledRanges.js`](https://github.com/stylelint/stylelint/blob/f38bd78bbae75284ed54e43516c4c8aceafa06e9/lib/assignDisabledRanges.js#L144-L162), we `walk` through comments:

https://github.com/stylelint/stylelint/blob/f38bd78bbae75284ed54e43516c4c8aceafa06e9/lib/assignDisabledRanges.js#L32

https://github.com/stylelint/stylelint/blob/f38bd78bbae75284ed54e43516c4c8aceafa06e9/lib/assignDisabledRanges.js#L144-L162

However, comments _within selectors or values_ are not added to the collection of `comment` nodes e.g. the following CSS:

```css
multiline /* selector comment */
selector {
  property: 
    multiline /* value comment */
    value;
}
```

Produces [the following AST](https://astexplorer.net/#/gist/92e60137d38bb6bfac5bdef35911abc2/d873af7ac8516f2db65aa5e4e7f6d2a34012b1f2):

```json
{
  "type": "root",
  "nodes": [
    {
      "raws": {
        "before": "",
        "between": " ",
        "selector": {
          "value": "multiline \nselector",
          "raw": "multiline /* selector comment */\nselector"
        },
      },
      "type": "rule",
      "nodes": [
        {
          "raws": {
            "before": "\n  ",
            "between": ": \n    ",
            "value": {
              "value": "multiline \n    value",
              "raw": "multiline /* value comment */\n    value"
            }
          },
          "type": "decl",
          "prop": "property",
          "value": "multiline \n    value"
        }
      ],
      "selector": "multiline \nselector"
    }
  ]
}
```

Notice how there are no nodes of `type` `comment` and the comments are stripped from the value of the `value` properties of each node. _However, and I think this is a change introduced in PostCSS@6, the comment is available within the value of the `raw` property_. Therefore, it _might be possible (no guarantees_) to use this information in `assignDisabledRanges.js` to identify these comments within values & selectors and then act on them. Investigating this potential would be a great way to help out.

**Comment by fiskhandlarn** [comment]
@jeddy3 Whoa, this is a little above my head atm, and I won't have time to read up enough to fix this, sorry. X)

**Comment by fiskhandlarn** [comment]
Anyone? X) 

![ezgif-4-7a3aae8a8905](https://user-images.githubusercontent.com/680264/55712921-eb39cb80-59ef-11e9-96c5-df4a85f424b0.gif)

**Comment by fiskhandlarn** [comment]
@jeddy3 ? X)

**Comment by eedwards-sk** [comment]
Any updates on this? I cannot disable `string-quotes` unless I completely disable stylelint:

this works
```
	& .name {
		/* stylelint-disable */
		@mixin FontSizer 2;
		/* stylelint-enable */

		font-weight: 700;
		text-transform: lowercase;
	}
```

this doesn't
```
	& .name {
		/* stylelint-disable string-quotes */
		@mixin FontSizer 2;
		/* stylelint-enable */

		font-weight: 700;
		text-transform: lowercase;
	}
```

and this doesn't
```
	& .name {
		/* stylelint-disable-next-line string-quotes */
		@mixin FontSizer 2;

		font-weight: 700;
		text-transform: lowercase;
	}
```

and this doesn't
```
	& .name {
		@mixin FontSizer 2; /* stylelint-disable-line string-quotes */

		font-weight: 700;
		text-transform: lowercase;
	}
```

**Comment by eedwards-sk** [comment]
even this doesn't work:

```
	/* stylelint-disable */
	& .name {
		@mixin FontSizer 2;

		font-weight: 700;
		text-transform: lowercase;
	}
	/* stylelint-enable */
```

```
src/components/Location-address/styles.css
 1:9  ✖  Expected single quotes   string-quotes
```

Does this feature work at all?

**Comment by jeddy3** [comment]
The issue is labelled as "help wanted" and is looking for someone to champion it.

**Comment by romainmenke** [comment]
Because we have been improving support for disable comments, auto fixes, ... I wanted to also take a look at this issue.

I have a working prototype here: https://github.com/stylelint/stylelint/pull/7839

The current approach is:
- rewrite the helper functions so that their arguments no longer expect a `PostCSS` `Comment` node. Instead they can take any comment from any node type.
- look for comments inside the raws of selectors, at rule params, ...
- tokenize those raws with `@csstools/css-tokenizer`

This seems to work well.
The one drawback is that it doesn't support non-standard CSS given that `@csstools/css-tokenizer` only tokenizes standard CSS tokens.

This limitation only applies to comments inside selectors, values, ... things that didn't work before anyway. So the fix for this issue is limited to standard CSS.

Is this approach something we want to pursue further?

**Comment by Mouvedia** [comment]
> The one drawback is that it doesn't support non-standard CSS given that @csstools/css-tokenizer only tokenizes standard CSS tokens.
> This limitation only applies to comments inside selectors, values, ... things that didn't work before anyway.

This means it's not a regression but could be unexpected hence we will have to add a note/warning in the documentation even if we are not supporting less/sass comments anymore.

> Is this approach something we want to pursue further?

There are probably rules that expect the current behaviour that is mentioned in the documentation:

> Comments within selector and value lists are currently ignored.

Those cases will have to be found and potentially tests will have to be added.
i.e. tests weren't added because we thought it would never change (bad regression coverage)
e.g. maybe some rules are stripping comments

Also some autofixes might not keep the stylelint commands that are within a selector which have a far reaching scope.
e.g. `stylelint-disable`
Since this issue is about `stylelint-disable-line` do we ignore the other stylelint commands?
If not should we change the title of this issue?

**Comment by romainmenke** [comment]
> > The one drawback is that it doesn't support non-standard CSS given that @csstools/css-tokenizer only tokenizes standard CSS tokens.
This limitation only applies to comments inside selectors, values, ... things that didn't work before anyway.

> This means it's not a regression but could be unexpected hence we will have to add a note/warning in the documentation even if we are not supporting less/sass comments officially anymore.

Yes, we should be clear about this.

-----

> > Is this approach something we want to pursue further?

> There are probably rules that expect the current behaviour that is mentioned in the documentation:

With `this approach` I meant the technical approach I outlined.

I am assuming that we still want to solve this issue in general
If we don't have consensus that this is something that we want to fix, please let me know.

---------

> There are probably rules that expect the current behaviour that is mentioned in the documentation:

I don't see how rules could depend on the current behavior?
Can you provide an example or a failing test?

------

> > Comments within selector and value lists are currently ignored.

> Those cases will have to be found and potentially tests will have to be added.
i.e. tests weren't added because we thought it would never change (bad regression coverage)
e.g. maybe some rules are stripping comments

> Also some autofixes might not keep the stylelint commands that are within a selector which have a far reaching scope.
e.g. stylelint-disable should be kept for the potential attempted fixes that come after

This seems out of scope?

Rules that apply fixes and erase comments is separate from comments for disable ranges.
We don't have to fix all rules and ensure that they do not erase comments to resolve this issue.

Maybe best to track this aspect in a separate issue?

-----

> Since this issue is about stylelint-disable-line do we ignore the other stylelint commands?
> If not should we change the title of this issue?

This issue is about all disable commands.
Feel free to change the title :)

**Comment by Mouvedia** [comment]
> > Also some autofixes might not keep the stylelint commands that are within a selector which have a far reaching scope.
> > e.g. stylelint-disable should be kept for the potential attempted fixes that come after
> 
> This seems out of scope?
> 
> Rules that apply fixes and erase comments is separate from comments for disable ranges. We don't have to fix all rules and ensure that they do not erase comments to resolve this issue.
> 
> Maybe best to track this aspect in a separate issue?

Fixing this issue would have the direct consequence of rendering those rules buggy when they weren't before.
This needs to be fixed in the same version, in separate PRs if necessary.

Here's a [demo](https://stylelint.io/demo/#N4Igxg9gJgpiBcICGACYAdAdugLpANhAE7woASAygDICCAFFrrgPQBUKAzjgJ74z4BLTDgC0UARyQAjPiMwwAHqMHyUKVs0Y5cYJPjB0AbkiJ0RIgBYBXGAEoUAahQBmAAy2ANFtwAmABwApB4oAKyuQSiuAHQ+WrYA3FgAvoyY8JgQOAzYOEheOWycPHwqouKSMjByispCMGoaWgAiAJIAalj2wEkgHiAAZgJ8AHJIALZwiIrjAA58UWAcHL3gEJiDAOYIIBiYauggRFZ8HAeku2r7A1aYYDgCa2ISeoQA7jBQIoJcZygA2gddPoDgBdfKXFCAiCEIgifo3O4PTByTJIe5rX4HMbQGBEbC9LAQg4cfgwO7EEQ8GZVXQkzEgN64g7gy7E0nk2EZURctFI+kcARjOYwA6ElApTA9PqQdYCDYAMWIYzR2wAVhw1itYDNlogLpCQFxePwhNoEAb8GiYD8CXsDUaSqaRDLNiIuEhMFATFB6ZacNazckQEkgA) with a rule that keeps the comments and one that doesn't.
Currently the rule that strip the comments is removing user comments which could maybe be argued that they ought to be kept, but it's nothing major. Once these stylelint commands won't be considered bogus, the removal of those comments will have an impact that I don't consider out of scope.
i.e. suddenly the removal is not just lost documentation but something that require that all rules be reviewed and fixed

**Comment by romainmenke** [comment]
> Fixing this issue would have the direct consequence of rendering those rules buggy when they weren't before.

No, they would be equally buggy :)

For it to be a new bug these things all need to be true:
- someone has a problem that can be fixed
- their config allows the problem to be fixed
- they haven't run with `--fix` yet
- the problematic CSS contains a disable command
- removing the disable command creates a problem for the user

This is absolutely impossible.
Especially because of that last line :)
No user can have a disable command that currently doesn't have an effect but that will introduce a negative effect when removed.

------

Let's retain the scope of this issue and discuss fixes that remove comments separately.
It is a completely different issue and unrelated to this one.

Scope bleed makes it much harder to discuss and resolve these issues :)

**Comment by Mouvedia** [comment]
> > removing the disable command creates a problem for the user
> This is absolutely _impossible_.
> Especially because of that last line :)

You mean that you assume that no one would depend on those stylelint command in that context because they have read the documentation warning.
i.e. I disagree because currently a user could add the command and come to the conclusion that it works 
Hence it's not impossible that lines that should be ignored—for a reason that was foreign to the user—would suddenly be fixed.
e.g. for the rules that keep the comments after the fix

If you are assuming that all users are aware of that warning and knew that fixing used to be disabled when certain disable comments were present in the file, indeed you would be right.

My example use `stylelint-disable-next-line`, but imagine the consequences if it was `stylelint-disable`.
It would suddenly attempt fix all the rest of the lines.
The reason these lines were ignored before doesn't matter for the user.
i.e. his assumption was wrong but the consequences remains

Is it unlikely? Yes. But once this issue is fixed, it will be an acceptable bug.
So to answer

> > Is this approach something we want to pursue further?
> With this approach I meant the technical approach I outlined.

I think it should be fixed but it will require to add plenty of tests and be accompanied by other PRs that fixes the consequences.
i.e. it's a trivial/independent fix IMHO

**Comment by romainmenke** [comment]
Can we please keep this on topic @Mouvedia?

It makes issues much harder to follow and it also causes a lot of noise to subscribers of issues.

As I said before, please open a new issue if you think there are other bugs.

**Comment by Mouvedia** [comment]
> As I said before, please open a new issue if you think there are other bugs.

I will once this issue is fixed.
i.e. there are no bugs right now

**Comment by ybiquitous** [comment]
@romainmenke Thanks for trying to address this issue. Honestly, I'm just worried that comments parsed by PostCSS custom syntaxes like `postcss-scss` would not work, if we stopped using the `Comment` node... 🤔 

If possible, improving the PostCSS parser itself seems ideal to correctly get comment nodes within selectors or values... 😓

**Comment by Mouvedia** [comment]
@ybiquitous https://github.com/postcss/postcss/issues/1145#issuecomment-1397999466

**Comment by romainmenke** [comment]
@ybiquitous We will continue to use the `Comment` node wherever possible.
We also won't change anything about things that already work.

The proposed change only adds an extra step where we also check other parts of the AST for comments.

For example:

```css
.foo {
  color: rgb(255 /* stylelint-disable-line */ 0 0);
}
```

But I do share your concerns about other syntaxes.
This won't work because `// stylelint-disable-line` isn't a standard CSS comment.

```scss
.foo {
  color: rgb(
    255 // stylelint-disable-line
    0
    0
  );
}
```

**Comment by romainmenke** [comment]
It is fairly easy to add support for `//` comments in `@csstools/css-tokenizer`.
This would be gated behind a flag.

```js
	const t = tokenizer({
		css: 'a// a comment\nb',
		oneLineCommentsAllowed: true,
	});

	assert.deepEqual(
		collectTokens(t),
		[
			['ident-token', 'a', 0, 0, { value: 'a' }],
			['comment', '// a comment', 1, 12, undefined],
			['whitespace-token', '\n', 13, 13, undefined],
			['ident-token', 'b', 14, 14, { value: 'b' }],
			['EOF-token', '', -1, -1, undefined],
		],
	);
}
```

There is some CSS that contains `//` that is not a comment:

```css
.a {
  background: url(https://example.com);
  border-image: url("/images/border.png") 27 23 // 1rem round space;
}
```

Setting this flag to `true` would cause unexpected results.
But this doesn't matter when looking specifically for comments and disregarding the rest of the tokens.

You could find a comment with `//example.com`, but this doesn't contain `stylelint...` so it is ignored anyway.

-----

I think this could be a way forwards for this issue while also having good support for SCSS and Less.

But then someone needs to help me with the very complicated existing logic for SCSS and Less comments 😅.

**Comment by ybiquitous** [comment]
@romainmenke Thanks for describing the solution. If we can maintain backward compatibility with other syntaxes like SCSS, there will be no problems.

And for now, I don't think supporting non-standard comments like `//` is necessary. We can still use `/* */` comments on such non-standard syntaxes, and we should usually avoid introducing more complexity than needed. Thanks for considering a possible solution. Let's revisit it if people would strongly request `//`.

The direction of #7839 sounds good to me. 👍🏼
